Guide for application specific email templates. (#1882)
FusionAuth/fusionauth-site: site/docs/v1/tech/guides/configuring-application-specific-email-templates.adoc

[#list state!{} as key, value][#if key != "tenantId" && key != "client_id" && value??][#assign url = url + "&" + key?url + "=" + value?url/][/#if][/#list]

${url}

- FusionAuth Admin
```

Be sure to save your icon by clicking the blue save icon.

When a user clicks the `Forgot Password` link on the hosted login page or calls the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[forgot password API], this template will be used to email the user to start the workflow to change their password for all applications in the specified Tenant - so long as they don't have a template set at the Application level.

Make sure you've created an Application in the link:/docs/v1/tech/core-concepts/applications[Application overview]. After, log in to your application by clicking on `Applications` in the navigation sidebar, selecting the green magnifying glass icon, and copying and pasting the `Login URL` into an incognito browser. Click the Forgot Password link, enter your email in the form, and you should see the `Forgot Password` email sent to your inbox.

> === Overriding Email Templates using API

Reviewer: ```suggestion
=== Configuring Application Email Templates
```

> Finally, we'll override our generic Tenant Email Template we just configured with a custom application specific template for a fictitious company called Pied Piper.

Reviewer: Change `we` to `you`.

Navigate to [breadcrumb]#Customizations -> Emails Templates# and click the blue edit icon for the `Forgot Password` email template and copy the `HTML Template` `Text Template` sections into a notepad.

Create a new email template by going to [breadcrumb]#Customizations -> Emails Templates# and click the green + icon.
Fill out the `Name`, `Default Subject`, `From Email`, `Default from Name` fields.
Paste your default template from above into the `HTML Template`  and `Text Template` sections. Feel free to edit these templates based on your application.

Save your template by clicking the blue save icon in the top right corner.

=== Test the Custom Template

> For our demo application Pied Piper, we'll call the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow] for an existing user. If you haven't already created a user, go ahead and complete the process in the  link:/docs/v1/tech/tutorials/register-user-login-api[Register a User and Login] guide.

Reviewer: ```suggestion
For our demo application Pied Piper, we'll call the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow] for a user. You can also use the login page link, as above.
```

They couldn't get access to the admin ui without creating a user.

```shell
curl --request POST \

>   --url https://localhost/api/user/forgot-password \

Reviewer: ```suggestion
   https://YOUR_FUSIONAUTH_INSTANCE/api/user/forgot-password \
```

>   --header 'Authorization: xxxxxx' \

Reviewer: ```suggestion
  --header 'Authorization: YOUR_API_KEY' \
```

  --header 'Content-Type: application/json' \
  --data '{

>     "applicationId": "85a03867-dccf-4882-adde-1a79aeec50df",

Reviewer: Does the user need to be registered for the application? Will the email send (with the application specific template) if they are not?

Author: Added a note that the user needs to be registered for the email to be sent. 👍

    "loginId": "dinesh@piedpiper.com",
     "sendForgotPasswordEmail": true
}'
```

> Modify the curl command based on the values for your application and user.

Reviewer: ```suggestion
Modify the curl command for your application, API key, FusionAuth instance location, and user.
```

> Once successfully executed, you should see your custom email sent to the above user's email address.

Reviewer: ```suggestion
Once successfully executed, you will see your application specific email sent to the above user's email address.
```

=== Override Email Template using the Hosted Login Page

In addition to calling the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow] API, you can also append the query string parameter to your `Forgot Password` link on the FusionAuth link:/docs/v1/tech/core-concepts/integration-points#hosted-login-pages[hosted login screen].

> Grab your Login URL under [breadcrumb]#Applications -> click the green magnifying glass icon for your application -> copy a Login URL# and append `&applicationId=123`, replacing 123 with your [field]#applicationId#. Enter your email address and submit the form. You should again see your application email template for Forgot Password.

Reviewer: ```suggestion
Grab your Login URL under [breadcrumb]#Applications# by clicking the green magnifying glass icon for your application and copying the [uielement]#Login URL#. Append `&applicationId=YOUR_APPLICATION_ID`, replacing `YOUR_APPLICATION_ID` with your [field]#applicationId#. Enter your email address and submit the form. You should again see your Forgot Password application email content used .
```

Reviewer: Is it `applicationId` or `clientId` or both?

Author: When you append the `applicationId` to the Login URL both would be included but the clientId is already included in the Login URL.

> By providing the [field]#applicationId# to the  link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow], FusionAuth was able to resolve the application specific email template.

Reviewer: ```suggestion
By providing the [field]#applicationId# to the  link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow], FusionAuth delivered the application specific email template.
```

Make sure you've created an Application in the link:/docs/v1/tech/core-concepts/applications[Application overview]. After, log in to your application by clicking on `Applications` in the navigation sidebar, selecting the green magnifying glass icon, and copying and pasting the `Login URL` into an incognito browser. Click the Forgot Password link, enter your email in the form, and you should see the `Forgot Password` email sent to your inbox.

=== Configuring Application Email Templates

Finally, you'll override your generic Tenant Email Template you just configured with a custom application specific template for a fictitious company called Pied Piper.

Navigate to [breadcrumb]#Customizations -> Emails Templates# and click the blue edit icon for the `Forgot Password` email template and copy the `HTML Template` `Text Template` sections into a notepad.

Create a new email template by going to [breadcrumb]#Customizations -> Emails Templates# and click the green + icon.

> Fill out the `Name`, `Default Subject`, `From Email`, `Default from Name` fields.

Author: Okay, I am going to say `Name`, `Default Subject`, `From Email`, `Default from Name`, `Default HTML`, and `Default Text` using the [field][#oijdf#format#. Going to leave `HTML Template` and `Text Template` in backticks because they are technically links, not fields. @mooreds

Paste your default template from above into the `HTML Template`  and `Text Template` sections. Feel free to edit these templates based on your application.

Save your template by clicking the blue save icon in the top right corner.

Now configure the email template by navigating to [breadcrumb]#Applications -> Your Application -> click the green edit icon -> Email -> Templates -> Forgot password# and select your application email template - this option will have the same name you populated the `Name` field with from above. Click the blue save icon in the top right corner.

=== Test the Custom Template

For our demo application Pied Piper, we'll call the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow] for a user. You can also use the Login URL, as previously demonstrated.
```shell
curl --request POST \

>     https://YOUR_FUSIONAUTH_INSTANCE/api/user/forgot-password \

Reviewer: ```suggestion
    YOUR_FUSIONAUTH_INSTANCE/api/user/forgot-password \
```

If running locally, they will use `http` not `https.

  --header 'Authorization: YOUR_API_KEY' \
  --header 'Content-Type: application/json' \
  --data '{
    "applicationId": "85a03867-dccf-4882-adde-1a79aeec50df",
    "loginId": "dinesh@piedpiper.com",
     "sendForgotPasswordEmail": true
}'
```

> Modify the curl command for your `applicationId`, `YOUR_API_KEY`, `YOUR_FUSIONAUTH_INSTANCE`, and `loginId` with the email address of the user you wish to send the email to. Remember this user must be registered for the application before the email can be sent.

Reviewer: ```suggestion
Modify the curl command. You'll need to update a number of values.

* Replace `YOUR_API_KEY` with a valid FusionAuth API key.
* Replace `YOUR_FUSIONAUTH_INSTANCE` with the URL to your instance. If running locally this will be `http://localhost:9011`.
* Replace the value of `applicationId` with the Id of your application.
* Replace the value of `loginId` with the email address of the account for which the password is being reset. Remember this user must be registered for the application before the email can be sent.
```

> Once successfully executed, you will see your application specific email sent to the above user's email address.

Reviewer: ```suggestion

Once successfully executed, you will see an application specific email sent to the above user's email address.
```

> === Override Email Template using the Hosted Login Page

Reviewer: We tend to call these "Hosted Login Pages", not singular. Any reason this is singular?

In addition to calling the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Start Forgot Password Workflow] API, you can also append the query string parameter to your `Forgot Password` link on the FusionAuth link:/docs/v1/tech/core-concepts/integration-points#hosted-login-pages[hosted login screen].

> Grab your Login URL under [breadcrumb]#Applications# by clicking the green magnifying glass icon for your application and copying the [uielement]#Login URL#. Append `&applicationId=YOUR_APPLICATION_ID` to your Login URL, replacing `YOUR_APPLICATION_ID` with your [field]#applicationId#. Enter your email address and submit the form. You should again see an application specific Forgot Password email in your user's inbox.

Reviewer: The login url? Don't you need to do something with the forgot password url? Did you mean to say "go to the login url page, then copy the forgot password link"? Or do I totally miss this?

I'd probably drop this whole section unless this is behavior we are willing to commit to supporting in the future.

Author: I tested both the Login URL and the Forgot Password URL and they both fire the application-specific email when the `applicationId` is provided. I believe this was part of the original purpose of the guide: if you provide an`applicationId` for action X for an API or a hosted page, we'll send an application-specific email if configured. I'll instruct them to append the `applicationId` to the Forgot Password URL. Since most folks use the hosted pages, I think it's important to include these instructions. However, if we just used the clientId to check if an application-specific email was configured, this would save the manual work, and provide a better user experience.

Author: Remove this section 👍

---
layout: doc
title: Configure Application Email Templates
description: I want to use application specific email templates to override Tenant level templates
navcategory: developer
---

:page-liquid:

:sectnumlevels: 0

include::docs/v1/tech/shared/_premium-edition-blurb.adoc[]

> 

Reviewer: Add a table of contents.

Reviewer: I use this command: `egrep '^[=]+ ' site/docs/v1/tech/PATH/TO/DOC |sed 's/=//' |sed 's/=/*/g'|sed 's/* /* <</'|sed 's/$/>>/'`

== Application Specific Email Templates

> Before you begin, complete the following guides:

Reviewer: Can we add a `== Prerequisites` section? Maybe the prereqs are:

* Stand up a FusionAuth instance
* Add an application with the authorization code grant enabled and at least one authorized redirect URL
* Configure an email provider

> Before you begin, complete the following guides:

Author: I think adding a Prerequisites section is a great idea, especially for new users/customers. 👍 @mooreds

> Before you begin, complete the following guides:

Author: I still the links should be provided to make it easier for users to quickly get to the relevant guides/tutorials 
`== Prerequisites

    * link:docs/v1/tech/getting-started/_5-minute-setup-guide.adoc[5 minute setup guide] - get a FusionAuth instance up and running
    * link:docs/v1/tech/email-templates/configure-email[Configure Email] - Configure an email provider
    * link:/docs/v1/tech/core-concepts/applications[Application overview] - Create an Application
    * link:docs/v1/tech/tutorials/register-user-login-api[Register a user and login] - Create  a User, register the User for your Application, and configure the Application for the Authorization Code Grant`

> Before you begin, complete the following guides:

Reviewer: okay, maybe something like:

```
== Prerequisites

In order for you to get the most value from this guide, you should have a FusionAuth instance, with an email provider and  an application set up for a user to log in with.

If you don't have this set up yet, please review the following links:

* link:/docs/v1/tech/getting-started/_5-minute-setup-guide.adoc[5 minute setup guide] - get a FusionAuth instance up and running with a simple application
* link:docs/v1/tech/email-templates/configure-email[Configure Email] - Configure an email provider. You can also use mailcatcher as a local SMTP client. This is link:/docs/v1/tech/installation-guide/docker#other-services[documented for a docker install].
```

The last two links are duplicative of the 5 minute guide.

    * link:docs/v1/tech/getting-started/_5-minute-setup-guide.adoc[5 minute setup guide]
    * link:docs/v1/tech/email-templates/configure-email[Configure Email]
    * link:/docs/v1/tech/core-concepts/applications[Application overview]
    * link:docs/v1/tech/tutorials/register-user-login-api[Register a user and login]

In FusionAuth, a Tenant can be configured to send transactional emails using default Email Templates that ship with FusionAuth, such as `Email Update`, `Forgot Password`, `Suspicious login`, etc. See the full list of email templates that can be configured here: link:docs/v1/tech/core-concepts/tenants#template-settings[Template Settings]

To begin, we'll configure our email template at the Tenant level to ensure they are sent when a user starts the link:docs/v1/tech/apis/users[Start Forgot Password Workflow].

> Next, navigate to [breadcrumb]#Tenant -> Email -> Template Settings -> Forgot Password#. Select `Forgot Password` from the dropdown, then click the blue save icon. When a user clicks the `Forgot Password` link on the hosted login page or calls the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[forgot password API], this template will be used to email the user to start the workflow to change their password for all applications in the specified Tenant - so long as they don't have a template set at the Application level.

Reviewer: So, if you are going against a new FusionAuth install, you won't see any templates here. You have to add them first, under customizations -> email templates. Can you please include that step, maybe copying from the email templates forgot password example?

![image](https://user-images.githubusercontent.com/91825/214630052-7ec73368-4425-4b2f-a48a-643711ed0142.png)

Author: This is odd. I just pulled in a fresh docker image and spun up a container and I had some default email templates at the Tenant level. I'll include a code snippet of the Forgot Password email template and directions for how to create it, just to be sure. @mooreds

> Make sure you've created an Application in the link:/docs/v1/tech/core-concepts/applications[Application overview] guide. After, log in to your application by clicking on `Applications` in the navigation sidebar, select the green magnifying glass icon, and copy and paste the `Login URL` into an incognito browser. Click the Forgot Password link, enter your email in the form, and you should see the `Forgot Password` email sent to your inbox.

Reviewer: The core concepts isn't a guide, really. 

```suggestion
Log in to your application by clicking on `Applications` in the navigation sidebar, selecting the green magnifying glass icon, and copying and pasting the `Login URL` into an incognito browser. 

Click the Forgot Password link, enter your email in the form, and you should see the `Forgot Password` email sent to your inbox.
```

> ### Overriding email templates

Reviewer: ```suggestion
== Overriding Email Templates
```

Finally, we'll override our generic Tenant Email Template we just configured with a custom application specific template for a fictitious company called Pied Piper.

> Navigate to [breadcrumb]#Customizations -> Emails Templates# and click the green + icon to create a new template.

Reviewer: I'd suggest just duplicating the previous template you added.

Fill out the `Name`, `Default Subject`, `From Email`, `Default from Name` fields.

> Paste the below code snippet into the `HTML Template` section and modify based on your application:

Reviewer: I might say "This template is modified to reference a certain application. Feel free to modify this example as needed."

```
[#setting url_escaping_charset="UTF-8"]
To change your password for your Pied Piper account, click on the following link.
<p>
[#-- The optional 'state' map provided on the Forgot Password API call is exposed in the template as 'state'.
If we have an application context, append the client_id to ensure the correct application theme when applicable.
--]
[#assign url = "http://localhost:9011/password/change/${changePasswordId}?client_id=${(application.oauthConfiguration.clientId)!''}&tenantId=${user.tenantId}" /]
[#list state!{} as key, value][#if key != "tenantId" && key != "client_id" && value??][#assign url = url + "&" + key?url + "=" + value?url/][/#if][/#list]
<a href="${url}">${url}</a>
</p>
- Pied Piper Admin

```

> and for `Text Template` section, we'll use:

Reviewer: Use 'you' not 'we'

```
[#setting url_escaping_charset="UTF-8"]
To change your password for your Pied Piper account, click on the following link.

[#-- The optional 'state' map provided on the Forgot Password API call is exposed in the template as 'state' --]
[#assign url = "https://local.fusionauth.io/password/change/${changePasswordId}?tenantId=${user.tenantId}" /]
[#list state!{} as key, value][#if key != "tenantId" && value??][#assign url = url + "&" + key?url + "=" + value?url/][/#if][/#list]

${url}

- Pied Piper Admin

```

Save your template by clicking the blue save icon in the top right corner.

> ### Test our custom template

Reviewer: ```suggestion
== Test the Custom Template
```

It's not markdown, so use `==`. Make sure the heading is titlecased.

> For our demo application Pied Piper, we'll call the link:/docs/v1/tech/apis/users#start-forgot-password-workflow[Forgot Password API] for an existing user. If you haven't already created a user, go ahead and complete the process in the  link:/docs/v1/tech/tutorials/register-user-login-api[Register a User and Login] guide.

Reviewer: Why wouldn't you go back to the UX the way you did before?

Author: `applicationId` is documented through our APIs but its not that apparent how to override the Tenant email templates. I can also include the option of using the hosted Login URL.

Author: Added a section for appending the `applicationId` to the Login URL. @mooreds

```shell
curl --request POST \
  --url https://localhost/api/user/forgot-password \
  --header 'Authorization: xxxxxx' \
  --header 'Content-Type: application/json' \
  --data '{
    "applicationId": "85a03867-dccf-4882-adde-1a79aeec50df",
    "loginId": "dinesh@piedpiper.com",
     "sendForgotPasswordEmail": true
}'
```

> Modfy the curl command based on the values for your application and user.

Reviewer: Spelling

Once successfully executed, you should see your custom email sent to the above user's email address.

> By providing the [field]#applicationId# to the  link:docs/v1/tech/apis/users[Start Forgot Password Workflow], FusionAuth was able to resolve the application specific email template.

Reviewer: Need full link to start password workflow.
